frontend/src/scenes/surveys/utils.ts

```diff
@@ -99,39 +101,107 @@ export type NPSBreakdown = {
     promoters: number
     passives: number
     detractors: number
+    score: string
 }
 
-export function calculateNpsBreakdown(surveyRatingResults: SurveyRatingResults[number]): NPSBreakdown | null {
-    // Validate input structure
-    if (!surveyRatingResults.data || surveyRatingResults.data.length !== 11) {
+// NPS calculation constants
+const NPS_SCALE_SIZE = 11 // 0-10 scale
+const NPS_PROMOTER_MIN = 9 // 9-10 are promoters
+const NPS_PASSIVE_MIN = 7 // 7-8 are passives. 0-6 are detractors but we don't need a variable for that.
+
+interface NPSRawData {
+    values: number[]
+    total: number
+}
+
+/**
+ * Extracts raw NPS data from processed survey data
+ */
+function extractNPSRawData(processedData: QuestionProcessedData): NPSRawData | null {
+    if (
+        !processedData?.data ||
+        processedData.type !== SurveyQuestionType.Rating ||
+        !Array.isArray(processedData.data) ||
+        processedData.data.length !== NPS_SCALE_SIZE
+    ) {
         return null
     }
 
-    if (surveyRatingResults.total === 0) {
-        return { total: 0, promoters: 0, passives: 0, detractors: 0 }
+    return {
+        values: processedData.data.map((item) => item.value),
+        total: processedData.totalResponses,
     }
+}
 
-    const PROMOTER_MIN = 9
-    const PASSIVE_MIN = 7
+/**
+ * Extracts raw NPS data from legacy survey rating results
+ */
+function extractNPSRawDataFromLegacy(surveyRatingResults: SurveyRatingResults[number]): NPSRawData | null {
+    if (!surveyRatingResults?.data || surveyRatingResults.data.length !== NPS_SCALE_SIZE) {
+        return null
+    }
 
-    const promoters = surveyRatingResults.data.slice(PROMOTER_MIN, 11).reduce((a, b) => a + b, 0)
-    const passives = surveyRatingResults.data.slice(PASSIVE_MIN, PROMOTER_MIN).reduce((a, b) => a + b, 0)
-    const detractors = surveyRatingResults.data.slice(0, PASSIVE_MIN).reduce((a, b) => a + b, 0)
-    return { total: surveyRatingResults.total, promoters, passives, detractors }
+    return {
+        values: surveyRatingResults.data,
+        total: surveyRatingResults.total,
+    }
 }
 
-export function calculateNpsScore(npsBreakdown: NPSBreakdown): number {
-    if (npsBreakdown.total === 0) {
-        return 0
+/**
+ * Core NPS calculation logic - works with raw data arrays
+ */
+function calculateNPSFromRawData(rawData: NPSRawData): NPSBreakdown {
```

Comment:
the `calculateNpsBreakdown` is sort of renamed to this function, with very slight changes just to add their score.

since the format is now different depending if it's the legacy viz or the new viz.

eventually, the older/legacy parts will be deleted